BUG: ndimage.median_filter: fix segfault when using `mode='mirror'`

aiudirog opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
https://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
https://scipy.github.io/devdocs/dev/contributor/development_workflow.html#writing-the-commit-message.
However, please only include an issue number in the description, not the title,
and please ensure that any code names containing underscores are enclosed in backticks.

Depending on your changes, you can skip CI operations and save time and energy: 
https://scipy.github.io/devdocs/dev/contributor/continuous_integration.html#skipping

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
Closes gh-22586

#### What does this implement/fix?
Fixed an off-by-one error in the 1D optimized implementation of `ndimage.median_filter()` when `mode='mirror'` is selected, which was resulting in segfaults down the line.

#### Additional information
- Added a test case that compares the 1D implementation to the general ND implementation
  - This test also reliably reproduced the segfault
- Updated the initialization of the working `data` array to use the C++ zerod syntax: `new T[len]();`

Please let me know if you need anything else!

aiudirog committed:
BUG:ndimage.median_filter: fix segfault when using the mode='mirror'

It appears there was an off-by-one error in the loop that finalizes the mirror mode. Additionally, added a test to directly compare the new 1D optimized implementation to the general ND one. Closes gh-22586

tylerjereddy commented on scipy/ndimage/tests/test_filters.py:
```diff
@@ -2766,3 +2766,40 @@ def test_gh_22333():
+        monkeypatch.setattr(
+            'scipy.ndimage._rank_filter_1d.rank_filter',
+            _1d_to_nd_rank_filter,
+        )
```
a bit convoluted--I guess there's no easy way to go through a public function directly to check?

aiudirog replied:
Yeah there isn't great access to get there without patching or adding another parameter. The closest I got was using an Nx1 array, but that didn't work for `mode='constant'` where it would just output a column of zeros.

tylerjereddy reviewed (commented):
I still need to test on my Linux box since the test passes even without the patch on my Mac.

aiudirog commented on scipy/ndimage/tests/test_filters.py:
```diff
@@ -2766,3 +2766,40 @@ def test_gh_22333():
+        )
+        desired = mf(arr)
+
+    np.testing.assert_array_almost_equal(actual, desired)
```
Ah, good catch, thanks!

aiudirog commented on scipy/ndimage/tests/test_filters.py:
```diff
@@ -2766,3 +2766,40 @@ def test_gh_22333():
+@pytest.mark.slow
+@pytest.mark.parametrize('length', range(1, 1000, 50))
+@pytest.mark.parametrize('size', range(1, 25))
+@pytest.mark.parametrize('mode', ['reflect', 'constant', 'nearest', 'mirror', 'wrap'])
```
Hypothesis was making a bit more difficult to find the cases that worked vs didn't. I'll rewrite the test now that it seems to be all good.

aiudirog commented on scipy/ndimage/tests/test_filters.py:
```diff
@@ -2766,3 +2766,40 @@ def test_gh_22333():
+    monkeypatch: pytest.MonkeyPatch,
+):
+    size = min(size, length)
+    arr = np.random.random(length)
```
This disappears in the hypothesis rewrite, but good to know for any future PRs. Thanks!

aiudirog committed:
Use assert_allclose instead of assert_array_almost_equal

aiudirog committed:
Refactored the test to use hypothesis

tylerjereddy commented:
I rebased and resolved merge conflicts

tylerjereddy commented on scipy/ndimage/tests/test_filters.py:
```diff
@@ -2982,3 +2986,44 @@ def test_edge_cases(self, xp):
+    size=st.integers(min_value=1, max_value=25),
+    mode=st.sampled_from(['reflect', 'constant', 'nearest', 'mirror', 'wrap']),
+)
+def test_gh_22586_median_filter_no_segfault_and_1d_matches_nd(
```
I'm not seeing a fail before/pass after application of the patch with this test in its current state on x86_64 Linux, even if I manually increase the 
 number of scenarios probed via `@settings(max_examples=5000)`.

Is anyone seeing this test fail on Linux if the source patch is reverted back to `main`?

In my hands, the original reproducing test from gh-22586 still does the trick. I suppose we could add both tests to be safe perhaps, but I don't even see a failure let alone a segfault for this test on `main` at the moment.

aiudirog replied:
I see the same thing and after some digging, it appears I may have made a mistake when converting the test to hypothesis? If you drop that commit, the parametrized pytest version segfaults if the changes to `_rank_filter_1d.cpp` aren't applied.

tylerjereddy commented:
We also don't usually add type hints to test functions, though I suppose it isn't wrong. In the interests of moving this forward for `1.15.3`, I may consider just pushing in the test that segfaults before/passes after as well, if I don't hear back by the weekend or so.

tylerjereddy committed:
TST: PR 22608 revisions

* Restore the original regression test that fails before
and passes after the pertinent patch.

tylerjereddy commented:
Yeah, I observed that too.

Also, now even with the patch applied I see a segfault for `spin test -t scipy/ndimage/tests/test_filters.py::test_gh_22586_crash_property` using `hypothesis` `6.131.9`. Sometimes it will pass but emit `double free or corruption (!prev)`. hmm...

aiudirog commented:
Doing the same on my system (also x86_64 Linux - AMD 9800X3D) tosses `corrupted size vs. prev_size` which is apparently a different heap corruption glibc message, so it would seem things are definitely still broken.

aiudirog commented on scipy/ndimage/src/_rank_filter_1d.cpp:
```diff
@@ -157,11 +157,7 @@ void _rank_filter(T *in_arr, int rank, int arr_len, int win_len, T *out_arr,
   int lim2 = arr_len - lim;
```
I think the other issue might be that this isn't validated to not be negative. Digging into it currently.

aiudirog replied:
Adding `if (lim2 < 0) return;` on the next line seems to stop the heap corruption. Do you see the same thing?

tylerjereddy committed:
BUG, MAINT: PR 22608 revisions

* Fix another memory issue in `_rank_filter()` based on feedback
at https://github.com/scipy/scipy/pull/22608#discussion_r2067771870.
This seems to allow many repeats of the property-based test to finally
pass on this branch via i.e.,
`spin test -t scipy/ndimage/tests/test_filters.py::test_gh_22586_crash_property -- --count=50`

* Also, remove unrelated whitespace change in `_rank_filter()`.

tylerjereddy commented:
I made a few changes here to see if we can have this ready for the `1.15.3` release:

1) I switched us back to the original property-based test used in the matching issue--it sounds like you were "ok" with that at https://github.com/scipy/scipy/pull/22608#issuecomment-2840537071.
2) I added your additional patch from https://github.com/scipy/scipy/pull/22608#discussion_r2067771870.
3) I removed an unrelated whitespace change in the diff.
4) I rebased on latest `main`.

With those changes, the following test incantation (50 repeats with `pytest-repeat`) reliably passes on this feature branch and fails (segfaults) when the source patch is reverted:

`spin test -t scipy/ndimage/tests/test_filters.py::test_gh_22586_crash_property -- --count=50`.

If the CI agrees/is happy, I'll probably go ahead and squash-merge for inclusion in `1.15.3`.

The main point of hesitation is that we maybe don't understand all the details of what is going on here, but that was also true of the previous patches applied on this code in `1.15.x` series, and preventing a segfault is good.

aiudirog commented:
Yeah this is good, definitely best to just make sure the segfault is gone.
My only concern is that it doesn't confirm that the 1D version gives the
same result as the ND version.

I did notice in cases where lim2 would be negative, the ND version does
produce an output different from the input, whereas this just does nothing.
Though I'm not sure if that result should be considered "undefined".

On Mon, May 5, 2025, 16:18 Tyler Reddy ***@***.***> wrote:

> *tylerjereddy* left a comment (scipy/scipy#22608)
> <https://github.com/scipy/scipy/pull/22608#issuecomment-2852233573>
>
> I made a few changes here to see if we can have this ready for the 1.15.3
> release:
>
>    1. I switched us back to the original property-based test used in the
>    matching issue--it sounds like you were "ok" with that at #22608
>    (comment)
>    <https://github.com/scipy/scipy/pull/22608#issuecomment-2840537071>.
>    2. I added your additional patch from #22608 (comment)
>    <https://github.com/scipy/scipy/pull/22608#discussion_r2067771870>.
>    3. I removed an unrelated whitespace change in the diff.
>    4. I rebased on latest main.
>
> With those changes, the following test incantation (50 repeats with
> pytest-repeat) reliably passes on this feature branch and fails
> (segfaults) when the source patch is reverted:
>
> spin test -t
> scipy/ndimage/tests/test_filters.py::test_gh_22586_crash_property --
> --count=50.
>
> If the CI agrees/is happy, I'll probably go ahead and squash-merge for
> inclusion in 1.15.3.
>
> The main point of hesitation is that we maybe don't understand all the
> details of what is going on here, but that was also true of the previous
> patches applied on this code in 1.15.x series, and preventing a segfault
> is good.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/scipy/scipy/pull/22608#issuecomment-2852233573>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABJGK76W2GNCRF4QOG4B6E3247BRPAVCNFSM6AAAAABYDB7XPOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDQNJSGIZTGNJXGM>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

tylerjereddy commented:
Ok, CI is happy--let's give this a try for `1.15.3` segfault guard.

I suspect this is just another step toward the actually-correct approach rather than perfecting things just yet.

tylerjereddy commented:
Thanks for your first contribution to SciPy
